src/java/picard/sam/RevertSam.java

```diff
@@ -148,76 +177,57 @@ public static void main(final String[] args) {
-                    allLibraryNamesIdentical = false;
-                }
-            }
-            if (SAMPLE_ALIAS != null && !allSampleAliasesIdentical) {
-                throw new PicardException("Read groups have multiple values for sample.  " +
-                        "A value for SAMPLE_ALIAS cannot be supplied.");
-            }
-            if (LIBRARY_NAME != null && !allLibraryNamesIdentical) {
-                throw new PicardException("Read groups have multiple values for library name.  " +
-                        "A value for library name cannot be supplied.");
-            }
-        }
+        ValidationUtil.validateHeaderOverrides(inHeader, SAMPLE_ALIAS, LIBRARY_NAME);
 
         ////////////////////////////////////////////////////////////////////////////
         // Build the output writer with an appropriate header based on the options
         ////////////////////////////////////////////////////////////////////////////
-        final boolean presorted = (inHeader.getSortOrder() == SORT_ORDER) || (SORT_ORDER == SortOrder.queryname && SANITIZE);
-        final SAMFileHeader outHeader = new SAMFileHeader();
-        for (final SAMReadGroupRecord rg : inHeader.getReadGroups()) {
-            if (SAMPLE_ALIAS != null) {
-                rg.setSample(SAMPLE_ALIAS);
-            }
-            if (LIBRARY_NAME != null) {
-                rg.setLibrary(LIBRARY_NAME);
-            }
-            outHeader.addReadGroup(rg);
-        }
-        outHeader.setSortOrder(SORT_ORDER);
-        if (!REMOVE_ALIGNMENT_INFORMATION) {
-            outHeader.setSequenceDictionary(inHeader.getSequenceDictionary());
-            outHeader.setProgramRecords(inHeader.getProgramRecords());
+        final boolean presorted = isPresorted(inHeader, SORT_ORDER, sanitizing);
+        if (SAMPLE_ALIAS != null) overwriteSample(inHeader.getReadGroups(), SAMPLE_ALIAS);
+        if (LIBRARY_NAME != null) overwriteLibrary(inHeader.getReadGroups(), LIBRARY_NAME);
+        final SAMFileHeader singleOutHeader = createOutHeader(inHeader, SORT_ORDER, REMOVE_ALIGNMENT_INFORMATION);
+        for (final SAMReadGroupRecord readGroup : inHeader.getReadGroups()) {
+            singleOutHeader.addReadGroup(readGroup);
         }
 
-        final SAMFileWriter out = new SAMFileWriterFactory().makeSAMOrBAMWriter(outHeader, presorted, OUTPUT);
+        final Map<String, File> outputMap;
+        final Map<String, SAMFileHeader> headerMap;
+        if (OUTPUT_BY_READGROUP) {
+            String defaultExtension = INPUT.toString().endsWith(".sam") ? ".sam" : ".bam";
+            outputMap = createOutputMap(OUTPUT_MAP, OUTPUT, defaultExtension, inHeader.getReadGroups());
+            ValidationUtil.assertAllReadGroupsMapped(outputMap, inHeader.getReadGroups());
+            headerMap = createHeaderMap(inHeader, SORT_ORDER, REMOVE_ALIGNMENT_INFORMATION);
+        } else {
+            outputMap = null;
```

Comment:
In general I like getting away from nulls, however I don't think Optional is helpful in this case. We should only ever try to access outputMap and headerMap if OUTPUT_BY_READGROUP is true. If we ever try to access them when OUTPUT_BY_READGROUP is false then there is a bug in the code somewhere, and it's appropriate to blow up in that case. I'd like to avoid masking the bug with a more obscure stack trace related to Optional, or silently succeeding but producing garbage.